feat: add discussion saviour bot as as lambda function

KhudaDad414 opened the pull request:
### Description

- This lambda function will work as the backend for the bot that we will use in slack to move discussions to GitHub.
- Before merging, we should make sure that the following environment variables are in place:
```
SLACK_TOKEN
GITHUB_TOKEN
DISCUSSION_TARGET_REPO_OWNER
DISCUSSION_TARGET_REPO_NAME
```

after this PR is merged, we can update its link in slack and everything should be in place.

### How it works
this is how a thread can be moved to GitHub:
1) Select `Save to GitHub` in the overflow menu of the message:

<img width="337" alt="Screenshot 2022-12-28 at 17 57 57" src="https://user-images.githubusercontent.com/32505158/209853657-be1068ca-8af3-4998-aeff-df3901b4334c.png">

2) Fill the title and select the discussion category that it belongs to:

<img width="531" alt="Screenshot 2022-12-28 at 17 58 21" src="https://user-images.githubusercontent.com/32505158/209853737-8a43c994-1b87-470e-87ad-a3bfef82564e.png">

3) Click on save.

4) A message will be posted in that thread confirming where the message is preserved.

<img width="815" alt="Screenshot 2023-02-23 at 18 25 21" src="https://user-images.githubusercontent.com/32505158/220997289-621f29cc-23a3-481b-945f-a8fedae4d9c2.png">

### How it looks

<img width="919" alt="Screenshot 2023-02-23 at 18 28 42" src="https://user-images.githubusercontent.com/32505158/220998099-2fbbb164-2db3-4426-9c5c-b7c56e17f76c.png">

KhudaDad414 committed:
feat: add discussion saviour bot as as lambda func

KhudaDad414 committed:
add dependency

KhudaDad414 committed:
final touches

magicmatatjahu commented on netlify/functions/save-discussion/index.ts:
```diff
@@ -0,0 +1,98 @@
+      statusCode: 400,
+    };
+  }
+  const payload = JSON.parse(
+    querystring.parse(event.body || '{}')?.payload as string
+  );
```
Why you have to parse query-string? Do we use any queries in request?

KhudaDad414 replied:
slack uses the `application/x-www-form-urlencoded` format. this was one of the ways of parsing it without any external library. any other suggestions?

magicmatatjahu replied:
Ok, thanks for clarification! Maybe add comment about that, because in the first look it's very weird 😆

magicmatatjahu commented on netlify/functions/save-discussion/index.ts:
```diff
@@ -0,0 +1,98 @@
+    discussion.parseReplies().catch((err) => {
+      axios.post(discussion.responseUrl, {
+        text: err.message,
+      });
```
You should also console.error any problems in this post, like:

```suggestion
      await axios.post(discussion.responseUrl, {
        text: err.message,
      }).catch(console.error);
```

magicmatatjahu commented on netlify/functions/save-discussion/index.ts:
```diff
@@ -0,0 +1,98 @@
+      payload.channel.id,
+      payload.response_url
+    );
+    discussion.parseReplies().catch((err) => {
```
```suggestion
    await discussion.parseReplies().catch((err) => {
```

magicmatatjahu commented on netlify/functions/save-discussion/index.ts:
```diff
@@ -0,0 +1,98 @@
+          'this discussion has been preserved here: ' + url
+        );
+      }
+    });
```
```suggestion
    const discussionUrl = await githubReposity.createDiscussion(discussion);
    if (url) {
      discussion.postMessage(
        'this discussion has been preserved here: ' + url
      );
    }
```

magicmatatjahu commented on netlify/functions/save-discussion/index.ts:
```diff
@@ -0,0 +1,98 @@
+    querystring.parse(event.body || '{}')?.payload as string
+  );
+  if (payload.type === 'message_action') {
+    console.log(payload);
```
```suggestion
```

magicmatatjahu commented on netlify/functions/save-discussion/GithubReposity.ts:
```diff
@@ -0,0 +1,124 @@
+    if (!discussion.replies) return true;
+    try {
+      for (const reply of discussion.replies) {
+        console.log('adding comment to discussion:', gitHubDiscussionId);
```
```suggestion
      console.log('adding comments to discussion:', gitHubDiscussionId);
      for (const reply of discussion.replies) {
```

magicmatatjahu commented on netlify/functions/save-discussion/GithubReposity.ts:
```diff
@@ -0,0 +1,124 @@
+        const commentId = addDiscussionComment.comment.id;
+        if (reply.isAnswer && commentId) {
+          this.markAnswer(commentId);
+        }
```
Can we mark comment as answer in the `addDiscussionComment` mutation or we really need separate call to the api? Maybe we can optimize it here 😄

magicmatatjahu replied:
Ok, we cannot :( https://docs.github.com/en/graphql/reference/input-objects#adddiscussioncommentinput

magicmatatjahu reviewed (commented):
Good work! As I see you didn't install the `@netlify/functions` dependency, so how did you test it 😄?

KhudaDad414 committed:
Update netlify/functions/save-discussion/index.ts

Co-authored-by: Maciej Urbańczyk <urbanczyk.maciej.95@gmail.com>

KhudaDad414 committed:
Update netlify/functions/save-discussion/index.ts

Co-authored-by: Maciej Urbańczyk <urbanczyk.maciej.95@gmail.com>

KhudaDad414 committed:
Update netlify/functions/save-discussion/index.ts

Co-authored-by: Maciej Urbańczyk <urbanczyk.maciej.95@gmail.com>

KhudaDad414 committed:
Update netlify/functions/save-discussion/index.ts

Co-authored-by: Maciej Urbańczyk <urbanczyk.maciej.95@gmail.com>

KhudaDad414 committed:
refactor the code

KhudaDad414 committed:
Update netlify/functions/save-discussion/GithubReposity.ts

Co-authored-by: Maciej Urbańczyk <urbanczyk.maciej.95@gmail.com>

KhudaDad414 committed:
add a comment

KhudaDad414 reviewed (commented):
@magicmatatjahu thanks for the review. I was actually testing it in another repo 😆
👁️ 🙏

KhudaDad414 committed:
Merge branch 'master' into add-bot-lambda

KhudaDad414 commented:
@derberg @magicmatatjahu @akshatnema pinging for visibility.

KhudaDad414 committed:
Merge branch 'master' into add-bot-lambda

akshatnema commented on package.json:
```diff
@@ -90,9 +90,11 @@
+    "@slack/web-api": "^6.5.1"
   },
   "devDependencies": {
+    "@netlify/functions": "^1.4.0",
```
Why we add `@netlify/functions` as a devDependency as we need netlify cli to integrate Netlify functions with frontend?

KhudaDad414 replied:
I have only been using some types from `@netlify/functions` and it's cli for testing functions. that's why I think it should be in devDependencies.

akshatnema reviewed (commented):
@KhudaDad414 Can you please add comments for future reference so that everyone can understand the implementation and logic behind code.

Also, will the `Save to Github` feature be enabled to all users? If yes, the user has to integrate GitHub with Slack in order to save the Slack in a Discussion?

KhudaDad414 committed:
refactor

KhudaDad414 committed:
Merge branch 'master' into add-bot-lambda

KhudaDad414 commented:
@akshatnema Hey mate. I have refactored the code and added comments for clarity. 👀 🙏 

> will the Save to Github feature be enabled to all users? If yes, the user has to integrate GitHub with Slack in order to save the Slack in a Discussion?

Yes, It will be enabled for all users. And no, users won't need a GitHub with Slack integration for this to work since we will use one of the bot's API keys to access Slack discussions and store them in GitHub.

KhudaDad414 committed:
Merge branch 'master' into add-bot-lambda

akshatnema commented:
Sorry for the delay @KhudaDad414, I was quite on medical leave and just recovered from it in weekend.

> Yes, It will be enabled for all users. And no, users won't need a GitHub with Slack integration for this to work since we will use one of the bot's API keys to access Slack discussions and store them in GitHub.

So, if everyone has access, it may happen the someone from the community spam a lot of threads in the GitHub discussions. We in this case should have control over this. And also, our API Key has been used to generate Github discussions, the token may result in rate limit due to excessive use by community members. So, in my opinion, you should have some restrictions on the use of this feature. WDYT? @derberg

KhudaDad414 committed:
Merge branch 'master' into add-bot-lambda

KhudaDad414 commented:
@akshatnema Happy to hear that you have recovered. Welcome back, mate!

KhudaDad414 commented:
> So, if everyone has access, it may happen the someone from the community spam a lot of threads in the GitHub discussions. We in this case should have control over this.

I think the best we can do is improve [the message](https://github.com/asyncapi/website/pull/1194/files#diff-fee5f1edc6f79c88ddb3cce28cef36ddc21f5ebfc5ec277ef8ffcec5294a241fR85) to show the user who saved the discussion and then we can block the bad actors from the Slack community itself.
or as you said limit the users who can use it, maybe TSC members only? would like to know what others think.

>  the token may result in a rate limit due to excessive use by community members.

I don't think the usage area of this feature is that common. Yes, it can happen but the user has to save around 100 discussions to use 1000 points. which I don't think is that worrisome.

akshatnema commented:
> or as you said limit the users who can use it, maybe TSC members only? would like to know what others think.

I will prefer for having this permission for TSC_Members only because for any thread in Slack, a TSC member can very well judge/decide whether this thread should be saved in discussion or not. Secondly, if a non-TSC member wants to save a thread of Slack, he can surely ask any of the TSC members or Community managers (like @derberg) to add the respective thread on GitHub. This is what I think about the functionality we should have.

derberg commented:
tbh I do not worry people will spam us because of misuse. We were afraid the same will happen with feedback for on docs, and yeah, nothing bad happened.

I'm more concerned on the limit of free functions invocation, but again, YOLO 😆 , if we will hit the limit, we can think about custom backend on digital ocean 🤷🏼 

In github discussion, at the top, in main message we need to indicate where it is coming from, and provide link to slack

btw Khuda you plan to write an article about it, right? RIGHT? 😄 

---

btw nice code, if only not TS 🤣

KhudaDad414 commented:
> In github discussion, at the top, in main message we need to indicate where it is coming from, and provide link to slack

It was there in the first version, 🤦  adding it now.

> btw Khuda you plan to write an article about it, right? RIGHT? 😄

right. 😄 

> btw nice code, if only not TS 🤣

Come on man, don't do whatever you did to Twitter to TS as well. 😆 some of us actually like it. 😄

KhudaDad414 committed:
final touches

KhudaDad414 committed:
Merge branch 'master' into add-bot-lambda

KhudaDad414 commented:
@derberg 👁️ 🙏

KhudaDad414 committed:
Merge branch 'master' into add-bot-lambda

akshatnema reviewed (approved):
I wish I could test this on some of my test repos but looks like will not be able to. But yeah, I believe in @KhudaDad414 and hence approving :heavy_check_mark:

akshatnema committed:
Merge branch 'master' into add-bot-lambda

KhudaDad414 commented:
@akshatnema thanks mate.

derberg commented:
@akshatnema yeah this dude already pushed some not tested changes to CI and I bet he learned to test things now before merging 😆 

@KhudaDad414 from Poland with love 😆 

btw netlify environment variable is set and ready

KhudaDad414 commented:
@derberg Those are some big allegations you are making brother.  (please don't send any links in this PR, we don't like that. 😆)

KhudaDad414 committed:
Merge branch 'master' into add-bot-lambda

KhudaDad414 commented:
/rtm
